Beta documentation for Snowflake Iceberg tables (#6162)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/snowflake-configs.md

- use the reference doc structure for this article / split into separate articles
--->

## Iceberg table format <Lifecycle status="beta"/>

> Snowflake Iceberg tables support for the dbt-snowflake adapter is available for three out-of-the-box materializations:  

Reviewer: maybe a small little intro of iceberg would be beneficial for users? e..g:

Snowflake Iceberg helps manage large datasets in Snowflake efficiently. Snowflake Iceberg tables support the dbt-snowflake adapter, which you can use for three out-of-the-box materializations:

> Snowflake Iceberg tables support for the dbt-snowflake adapter is available for three out-of-the-box materializations:  

Reviewer: maybe a small little intro of iceberg would be beneficial for users? e..g:

```suggestion
Snowflake Iceberg helps manage large datasets in Snowflake efficiently. Snowflake Iceberg tables support the dbt-snowflake adapter, which you can use for three out-of-the-box materializations:
```

Author: We're going to have a whole page we'll link to eventually. THis is just the config ref page docs for now.

- [Table](/docs/build/materializations#table)

> - [Incremental](/docs/build/materializations#view)

Reviewer: ```suggestion
- [Incremental](/docs/build/materializations#incremental)
```

- [Dynamic](#dynamic-tables) 

For now, Iceberg tables require a [behavior flag](/reference/global-configs/behavior-changes) due to performance impact related to using Iceberg tables. Snowflake does not support `is_iceberg` on the  `Show Objects` queries, which dbt depends on for metadata.

> To use Iceberg, set the Iceberg Materialization Behavior flag to true in your dbt_project.yml:

Reviewer: ```suggestion
To use Iceberg, set the Iceberg Materialization Behavior flag to `True` in your dbt_project.yml:
```

> To use Iceberg, set the Iceberg Materialization Behavior flag to `True` in your dbt_project.yml:

Reviewer: is "Iceberg Materialization Behavior flag" going to be the new doc name for this behavior flag page? only asking bc it doesn't match that actual flag `enable_iceberg_materializations`.

Author: Not sure yet. @amychen1776 Thoughts?

Reviewer 2: Two thoughts:
1. @matthewshaver we need a page for this specific behavior flag under Snowflake (related to your other PR)
2. let's update this to say set the `enable_iceberg_materializations` behavior flag to 'True'

<File name='dbt_project.yml'>

```yaml

flags:
  enable_iceberg_materializations: True

```

</File>

> If you do not set this behavior flag, you will get an error, and dbt won’t materialize the table in the Iceberg format.

Reviewer: ```suggestion
If you don't set this behavior flag, you will get an error and dbt won’t materialize the table in the Iceberg format.
```

The following configurations are supported: 

| Field | Type | Required | Description | Sample input |
|-------|------|----------|-------------|-------|
|Table Format | String | Yes | Configures the objects table format. | `iceberg` |

> |External volumne | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |

Reviewer: ```suggestion
|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |
```

> |Base location |String | No | Defines the directory path for Snowflake to write the table data and metadata files. You can’t change this after creating the table. If not specified, Snowflake will use the current schema. | `jaffle_marketing_folder` |

Reviewer 3: I think this makes more sense to me, but still doesn't make sense to why the word "schema" was used...

```suggestion
|Base location |String | No | Defines the directory path for Snowflake to write the table data and metadata files. You can’t change this after creating the table. If not specified, dbt-snowflake will chose one based on the model's current schema. | `jaffle_marketing_folder` |
```

> Example of an Iceberg table materialization configuration:

Reviewer: sugg adding an h3 for this example config
```suggestion
### Example configuration
To configure an Iceberg table materialization in dbt, refer to the example configuration:
```

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

> There are some limitations to the implementation you need to be aware of:

Reviewer: also sugg an h3 to separate this content
```suggestion
### Limitations
There are some limitations to the implementation you need to be aware of:
```

> - You cannot create transient or temporary Iceberg tables on Snowflake. 

Reviewer 3: why does this matter to end users? perhaps this is a better phrasing?

```suggestion
- Using Iceberg tables with dbt, the end result is that your query is materialized in Iceberg. However, as has always been the case in dbt, often intermediary objects such as temporary and transient tables are made at materialization time. It is not possible to configure these temporary objects to also be Iceberg-formatted.
```

Author: @amychen1776 Anything to add here?

Reviewer 2: Let's tweak this sentence a little bit:

>However, as has always been the case in dbt, often intermediary objects such as temporary and transient tables are made at materialization time.

To:

However, often times dbt creates intermediary objects as temporary and transient tables for certain materializations like incremental. It is not possible to configure these temporary objects also to be Iceberg-formatted. You may see non-Iceberg tables created in the logs to support specific materializations but they will be dropped after usage.

> - Supplying an input to `base_location_subpath` will always be apprehended to your schema name. Currently, you cannot override this behavior, which ensures that dbt can differentiate Iceberg model builds based on the environment. 

Reviewer: i think we mean appended, right?
```suggestion
- The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.
```

> - Snowflake has limitations for Dynamic Tables. Check out the [Snowflake docs]( https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake#usage-notes) for more information. By default, we recommend leaving the `base_location_subpath` field blank, as each target has it’s own default path.

Reviewer 2: ```suggestion
- By default, we recommend leaving the `base_location_subpath` field blank, as each target has it’s own default path.
```

Reviewer 2: I'm debating keeping this in:
>Snowflake has limitations for Dynamic Tables. Check out the [Snowflake docs]( https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake#usage-notes) for more information.

On one hand - this is useful to see but this just seems a little random. Maybe reword it to call out that Snowflake has limitations on Iceberg Table Format Dynamic Tables?

Reviewer 3: I don't know this bullet caveating usage of Dynamic Iceberg tables points to the generic Create Iceberg Table usage notes? what's the point being made here?

also, what does this mean?
> By default, we recommend leaving the `base_location_subpath` field blank, as each target has it’s own default path.

Reviewer 2: 1. Yeah let's just remove the bullet about the Dynamic table limitations. Users can google it themselves
2. And in terms fo base_location_subpath. That's a field that is optional and I think most users will leave it blank. @dataders I'm not quite sure what your confusion is - could you expand?

> - Copying rows from an existing default-format model into an Iceberg-format model is not currently supported. This means for incremental models, you will have to use `--full-refresh` to drop and recreate the table from scratch. An exception is thrown without `--full-refresh`.

Reviewer 3: ```suggestion
- You cannot incrementally update a preexisting incremental model to be an Iceberg table. To do so, you must fully rebuild the table with the `--full-refresh` flag.
```

> - Only `on_schema_change` strategies fail/ignore are supported at p0.

Reviewer 2: ```suggestion
- Only `on_schema_change` strategies fail/ignore are supported at this time. 
```

Reviewer 3: ```suggestion
- For `on_schema_change`, only the `fail` and `ignore` strategies are currently supported.
```

- use the reference doc structure for this article / split into separate articles
--->

## Iceberg table format <Lifecycle status="beta"/>

Snowflake Iceberg tables support for the dbt-snowflake adapter is available for three out-of-the-box materializations:  
- [Table](/docs/build/materializations#table)
- [Incremental](/docs/build/materializations#incremental)
- [Dynamic](#dynamic-tables) 

For now, Iceberg tables require a [behavior flag](/reference/global-configs/behavior-changes) due to performance impact related to using Iceberg tables. Snowflake does not support `is_iceberg` on the  `Show Objects` queries, which dbt depends on for metadata.

> To use Iceberg, set the Iceberg Materialization Behavior flag to `True` in your dbt_project.yml:

Author: ```suggestion
To use Iceberg, set the `enable_iceberg_materializations` flag to `True` in your dbt_project.yml:
```

<File name='dbt_project.yml'>

```yaml

flags:
  enable_iceberg_materializations: True

```

</File>

> If you don't set this behavior flag, you will get an error and dbt won’t materialize the table in the Iceberg format.

Reviewer 2: ```suggestion
If you don't set this behavior flag, you will get an error and dbt will not materialize the table in the Iceberg format.
```

The following configurations are supported: 

| Field | Type | Required | Description | Sample input |
|-------|------|----------|-------------|-------|
|Table Format | String | Yes | Configures the objects table format. | `iceberg` |
|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |

> |Base location |String | No | Defines the directory path for Snowflake to write the table data and metadata files. You can’t change this after creating the table. If not specified, dbt-snowflake will chose one based on the model's current schema. | `jaffle_marketing_folder` |

Reviewer 2: ```suggestion
|Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. If not specified, dbt-snowflake will default to one based on the model's current schema. | `jaffle_marketing_folder` |
```

Reviewer 4: > You can’t change this after creating the table.

I don't think this is precisely true. Like, you can. It'll just be a different table in your external storage. Technically, the old one is still there? Not sure how you want to capture that nuance in wording if at all.

Reviewer 2: @matthewshaver could we capture this in the limitations? If they rename the base_location_subpath, we will technically be creating a new table in the storage location.

### Example configuration
To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Limitations
There are some limitations to the implementation you need to be aware of:

> - You cannot create transient or temporary Iceberg tables on Snowflake. 

Author: ```suggestion
-  Using Iceberg tables with dbt, the result is that your query is materialized in Iceberg. However, often, dbt creates intermediary objects as temporary and transient tables for certain materializations, such as incremental ones. It is not possible to configure these temporary objects also to be Iceberg-formatted. You may see non-Iceberg tables created in the logs to support specific materializations, but they will be dropped after usage.
```

- The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

> - Snowflake has limitations for Dynamic Tables. Check out the [Snowflake docs]( https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake#usage-notes) for more information. By default, we recommend leaving the `base_location_subpath` field blank, as each target has it’s own default path.

Author: ```suggestion
- By default, we recommend leaving the `base_location_subpath` field blank, as each target has a default path.
```

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Limitations
There are some limitations to the implementation you need to be aware of:

- You cannot create transient or temporary Iceberg tables on Snowflake. 
- The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.
- Snowflake has limitations for Dynamic Tables. Check out the [Snowflake docs]( https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake#usage-notes) for more information. By default, we recommend leaving the `base_location_subpath` field blank, as each target has it’s own default path.
- Copying rows from an existing default-format model into an Iceberg-format model is not currently supported. This means for incremental models, you will have to use `--full-refresh` to drop and recreate the table from scratch. An exception is thrown without `--full-refresh`.

> - For `on_schema_change`, only the `fail` and `ignore` strategies are currently supported.

Reviewer 4: Correction: all 4 are supported!! We're shipping a PR for this today.

Reviewer 3: @VersusFacit so we can delete the caveat entirely?
```suggestion
```

Reviewer 3: @VersusFacit so we can delete the caveat entirely?
```suggestion
```

Reviewer 4: I like the idea of eliding redundant info! Helps with with readability.

- use the reference doc structure for this article / split into separate articles
--->

## Iceberg table format <Lifecycle status="beta"/>

> Snowflake Iceberg tables support for the dbt-snowflake adapter is available for three out-of-the-box materializations:

Reviewer 2: ```suggestion
Iceberg table format support on the dbt-snowflake adapter is available for three out-of-the-box materializations:
```

This sentence feels really awkward when you read it out loud. I tried to correct it but I don't think I'm quite there yet.

- [Table](/docs/build/materializations#table)
- [Incremental](/docs/build/materializations#incremental)
- [Dynamic](#dynamic-tables) 

> For now, Iceberg tables require a [behavior flag](/reference/global-configs/behavior-changes) due to performance impact related to using Iceberg tables. Snowflake does not support `is_iceberg` on the  `Show Objects` queries, which dbt depends on for metadata.

Reviewer 2: ```suggestion
For now, in order to create Iceberg tables, users must implement a [behavior flag](/reference/global-configs/behavior-changes) due to performance impact related to using Iceberg tables. Snowflake does not support `is_iceberg` on the  `Show Objects` query, which dbt depends on for metadata.
```

Author: How does this work for you?
```suggestion
For now, to create Iceberg tables, you must implement a [behavior flag](/reference/global-configs/behavior-changes) due to performance impact related to using Iceberg tables. Snowflake does not support `is_iceberg` on the `Show Objects` query, which dbt depends on for metadata.
```

> To use Iceberg, set the `enable_iceberg_materializations` flag to `True` in your dbt_project.yml:

Reviewer 3: 1. this feels like it should be a callout, but don't know if you can embed code blocks within an admonition
2. we should also update the behavior flags doc to include this new `enable_iceberg_materializations` flag

```suggestion
Today, to use Iceberg tables, you need to enable the `enable_iceberg_materializations`  [behavior flag](/reference/global-configs/behavior-changes).
The reason is there a potential performance impact because Snowflake does not return an `is_iceberg` column with `SHOW OBJECTS` queries, upon which dbt heavily depends for metadata.

If you don't set this behavior flag, you will get an error and dbt will not materialize the table in the Iceberg format.
```

Reviewer 2: I think there's a typo here - it should say Snowflake does not return

Reviewer 2: We will have a Snowflake specific adapter behavior page -> just waiting for that PR to be merged in that changes the alignment (at least I think it's still pending)

Reviewer 3: fixed the typo in the above suggestion

<File name='dbt_project.yml'>

```yaml

flags:
  enable_iceberg_materializations: True

```

</File>

> If you don't set this behavior flag, you will get an error and dbt will not materialize the table in the Iceberg format.

Reviewer 3: moved to before the code block
```suggestion
```

The following configurations are supported: 

| Field | Type | Required | Description | Sample input |
|-------|------|----------|-------------|-------|

> |Table Format | String | Yes | Configures the objects table format. | `iceberg` |

Reviewer 2: ```suggestion
|Table Format | String | Yes | Configures the object table format. | `iceberg` |
```

|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |

> |Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. If not specified, dbt-snowflake will default to one based on the model's current schema. | `jaffle_marketing_folder` |

Reviewer 2: ```suggestion
|Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. | `jaffle_marketing_folder` |
```

> |Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. If not specified, dbt-snowflake will default to one based on the model's current schema. | `jaffle_marketing_folder` |

Reviewer 3: imho, this table has too many words that are better described elsewhere. To simplify, I suggest:
1. linking to Snowflake's docs page rather than restate their descriptions
2. adding a `Note` column to improve readability and reference elsewhere

### current version

![image](https://github.com/user-attachments/assets/acbe3b99-b772-4939-b98b-c1acb39db8e9)

### render of table with my updates

![image](https://github.com/user-attachments/assets/f325fba1-4955-4196-9446-3456f62cba09)

```suggestion
The following configurations are supported.
For more information, see the Snowflake reference for [`CREATE ICEBERG TABLE` (Snowflake as the catalog)](https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake)

| Field                 | Type   | Required | Description                                                                                                                | Sample input              | Note                                                                                                                                                                                                                                                         |
| --------------------- | ------ | -------- | -------------------------------------------------------------------------------------------------------------------------- | ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Table Format          | String | Yes      | Configures the objects table format.                                                                                       | `iceberg`                 | `iceberg` is the only accepted value                                                                                                                                                                                                                          |
| External volume       | String | Yes(*)   | Specifies the identifier (name) of the external volume where Snowflake writes the Iceberg table's metadata and data files. | `my_s3_bucket`            | *You need not specify this if the account, database, or schema already has an associated external volume. [More info](https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table-snowflake#:~:text=Snowflake%20Table%20Structures.-,external_volume) |
| Base location Subpath | String | No       | An optional suffix to add to the `base_location` path that dbt automatically specifies                                     | `jaffle_marketing_folder` | The recommendation is that you do not specify this. Modifying this parameter results in a new Iceberg table. More info, see [Base Location](#base-location).                                                                                                  |
```

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",

>     base_location_subpath="jaffle_shop_folder",

Reviewer 3: if we don't want users configuring `base_location_subpath`, we shouldn't include it in an example

```suggestion
    base_location_subpath="jaffle_shop_folder",
```

  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

> If you were to create an Iceberg table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt abstracts away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:

Reviewer 3: clarified?

```suggestion
Snowflake's `CREATE ICEBERG TABLE` DDL requires that a `base_location` be provided. dbt defines this parameter on behalf of the user to both streamline usage and enforce basic isolation of table data within the `EXTERNAL VOLUME`. The default behavior in dbt is to provide a `base_location` string that follows the below convention:
```

`_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

> However, dbt allows users to configure a `base_location_subpath` that is empty by default but, when provided, will be concatenated to the end of the previously described pattern for `base_location` string generation. We recommend leaving the `base_location_subpath` field blank, as each target has a default path. The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

Reviewer 2: ```suggestion
However, dbt allows users to configure a `base_location_subpath` that is empty by default but, when provided, will be concatenated to the end of the previously described pattern for `base_location` string generation. We recommend leaving the `base_location_subpath` field blank by default. The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.
```

> dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the external volume declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:

Reviewer 2: ```suggestion
dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:
```

- Navigate the underlying object store (S3/Azure blob)
- Read Iceberg tables via an object-store integration
- Grant schema-specific access to tables via object store
- Use a crawler pointed at the tables within the external volume to build a new catalog with another tool

> In summary, dbt-snowflake does not support arbitrary definition of base_location for Iceberg tables. Instead, dbt, by default, writes your tables within a `_dbt/{SCHEMA_NAME}` prefix to ensure easier object-store observability and auditability.

Reviewer 2: ```suggestion
To maintain best practices,  we enforce an input. Currently, we do not support overriding the default `base location` input but will consider it based on user feedback. 

In summary, dbt-snowflake does not support arbitrary definition of base_location for Iceberg tables. Instead, dbt, by default, writes your tables within a `_dbt/{SCHEMA_NAME}` prefix to ensure easier object-store observability and auditability.
```

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

If you were to create an Iceberg table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt abstracts away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:

`_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

However, dbt allows users to configure a `base_location_subpath` that is empty by default but, when provided, will be concatenated to the end of the previously described pattern for `base_location` string generation. We recommend leaving the `base_location_subpath` field blank by default. The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

> dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:

Author: #### Why use the dbt generated `base_location`?

dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog keeps track of your Iceberg table regardless of where the data lives within the `external volume` declared and the `base_location` provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This behavior could result in future technical debt because it will limit the ability to:

> dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:

Author: ```suggestion
#### Why use the dbt generated `base_location`?

dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog keeps track of your Iceberg table regardless of where the data lives within the `external volume` declared and the `base_location` provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This behavior could result in future technical debt because it will limit the ability to:
```

Reviewer 2: That title sounds a little funny - they don't really have a choice if they want to use this config :)

Reviewer 3: @matthewshaver is this comment you editing my proposed comment? can I propose that you first accept my changes then do a revision?

rationale:
1. i get credit (and future blame)
2. it's easier to see what your edits are?

- use the reference doc structure for this article / split into separate articles
--->

## Iceberg table format <Lifecycle status="beta"/>

> Snowflake Iceberg tables support for the dbt-snowflake adapter is available for three out-of-the-box materializations:

Author: What if we split this into two sentences? Something like
```suggestion
The dbt-snowflake adapter supports the Iceberg table format. It is available for three of the Snowflake materializations: 
```

|Table Format | String | Yes | Configures the objects table format. | `iceberg` |
|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |
|Base location |String | No | Defines the directory path for Snowflake to write the table data and metadata files. You can’t change this after creating the table. If not specified, dbt-snowflake will chose one based on the model's current schema. | `jaffle_marketing_folder` |

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

> If you were to create an Iceberg Table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt Labs has chosen to abstract away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:

Reviewer 2: ```suggestion
If you were to create an Iceberg Table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt abstracts away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:
```

`_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

> However, dbt allows users to configure a `base_location_subpath` that is empty by default but, when provided, will be concatenated to the end of the previously described pattern for `base_location` string generation. We recommend leaving the `base_location_subpath` field blank by default. The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

Reviewer 3: didn't mean to edit as heavily as I did but here we are.

note i completely cut this line:
> This ensures dbt can differentiate Iceberg model builds by environment.

```suggestion
#### Base Location Subpath
We recommend using dbt's auto-generated `base_location`. However, if you have a need to customize the resulting `base_location`, dbt allows users to configure a `base_location_subpath`. When specified, it will be concatenated to the end of the previously described pattern for `base_location` string generation.

For example, `config(base_location_subpath="prod")` will generate a `base_location` of the form `_dbt/{SCHEMA_NAME}/{MODEL_NAME}`

A theoretical (not recommended) use case: re-using an `EXTERNAL VOLUME` while still maintaining isolation  across development and production environments. This is not recommended as storage permissions should configured on the external volume and underlying storage, not paths that can be modified by any analytics engineer.
```

> dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the external volume declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible. An empty string could result in future technical debt because it will limit the ability to:

Reviewer 2: ```suggestion
dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the external volume declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:
```

> dbt does this to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The base location declares where to write in the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, passing anything into the `base_location` field, including an empty string, is possible.  If a project does not enforce an input and always allows empty strings, this could result in future technical debt because it will limit the ability to:

Reviewer 3: ```suggestion
#### rationale

dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This could result in future technical debt because it will limit the ability to:
```

- Navigate the underlying object store (S3/Azure blob)
- Read Iceberg tables via an object-store integration
- Grant schema-specific access to tables via object store
- Use a crawler pointed at the tables within the external volume to build a new catalog with another tool

In summary, dbt-snowflake does not support arbitrary definition of base_location for Iceberg tables. Instead, dbt, by default, writes your tables within a `_dbt/{SCHEMA_NAME}` prefix to ensure easier object-store observability and auditability.

### Limitations

There are some limitations to the implementation you need to be aware of:

-  Using Iceberg tables with dbt, the result is that your query is materialized in Iceberg. However, often, dbt creates intermediary objects as temporary and transient tables for certain materializations, such as incremental ones. It is not possible to configure these temporary objects also to be Iceberg-formatted. You may see non-Iceberg tables created in the logs to support specific materializations, but they will be dropped after usage.

> - The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

Reviewer 2: ```suggestion
- The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.
```

We should move this up to the base location paragraph as well as the bullet underneath about base_location_subpath field

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
    base_location_subpath="jaffle_shop_folder",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

If you were to create an Iceberg table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt abstracts away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:

`_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

However, dbt allows users to configure a `base_location_subpath` that is empty by default but, when provided, will be concatenated to the end of the previously described pattern for `base_location` string generation. We recommend leaving the `base_location_subpath` field blank by default. The `base_location_subpath` input is always appended to your schema name, and this behavior cannot be overridden. This ensures dbt can differentiate Iceberg model builds by environment.

> #### rationale

Author: ```suggestion
#### Rationale
```

> dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This could result in future technical debt because it will limit the ability to:

Author: ```suggestion
#### Why use the dbt generated `base_location`?

dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog keeps track of your Iceberg table regardless of where the data lives within the `external volume` declared and the `base_location` provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This behavior could result in future technical debt because it will limit the ability to:
```

> dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog will keep track of your Iceberg table regardless of where the data lives within the `external volume` declared and what `base_location` is provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This could result in future technical debt because it will limit the ability to:

Author: ```suggestion
dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog keeps track of your Iceberg table regardless of where the data lives within the `external volume` declared and the `base_location` provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This behavior could result in future technical debt because it will limit the ability to:
```

|-------|------|----------|-------------|-------|
|Table Format | String | Yes | Configures the object table format. | `iceberg` |
|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |
|Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. | `jaffle_marketing_folder` |

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",
  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

> Snowflake's `CREATE ICEBERG TABLE` DDEL requires that a `base_location` be provided. dbt defines this parameter on behalf of the user to both streamline usage and enforce basic isolation of table data within the `EXTERNAL VOLUME`. The default behavior in dbt is to provide a `base_location` string that follows the below convention:

Author: ```suggestion
Snowflake's `CREATE ICEBERG TABLE` DDEL requires that a `base_location` be provided. dbt defines this parameter on the user's behalf to streamline usage and enforce basic isolation of table data within the `EXTERNAL VOLUME`. The default behavior in dbt is to provide a `base_location` string that follows the below convention:
```

> Snowflake's `CREATE ICEBERG TABLE` DDEL requires that a `base_location` be provided. dbt defines this parameter on the user's behalf to streamline usage and enforce basic isolation of table data within the `EXTERNAL VOLUME`. The default behavior in dbt is to provide a `base_location` string that follows the below convention:
> 
> `_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

Reviewer 3: 1. "DDL" not "DDEL"
2. in-lined the `base_location` string because it's not substantial to be it's own line
```suggestion
Snowflake's `CREATE ICEBERG TABLE` DDL requires that a `base_location` be provided. dbt defines this parameter on the user's behalf to streamline usage and enforce basic isolation of table data within the `EXTERNAL VOLUME`. The default behavior in dbt is to provide a `base_location` string of the form: `_dbt/{SCHEMA_NAME}/{MODEL_NAME}`
```

dbt manages `base_location` on behalf of users to enforce best practices. With Snowflake-managed Iceberg format tables, the user owns and maintains the data storage of the tables in an external storage solution (the declared `external volume`). The `base_ location` parameter declares where to write the data within the external volume. The Snowflake Iceberg catalog keeps track of your Iceberg table regardless of where the data lives within the `external volume` declared and the `base_location` provided. However, Snowflake permits passing anything into the `base_location` field, including an empty string, even reusing the same path across multiple tables. This behavior could result in future technical debt because it will limit the ability to:

- Navigate the underlying object store (S3/Azure blob)
- Read Iceberg tables via an object-store integration
- Grant schema-specific access to tables via object store
- Use a crawler pointed at the tables within the external volume to build a new catalog with another tool

To maintain best practices,  we enforce an input. Currently, we do not support overriding the default `base location` input but will consider it based on user feedback. 

> In summary, dbt-snowflake does not support arbitrary definition of base_location for Iceberg tables. Instead, dbt, by default, writes your tables within a `_dbt/{SCHEMA_NAME}` prefix to ensure easier object-store observability and auditability.

Reviewer 3: ```suggestion
In summary, dbt-snowflake does not support arbitrary definition of `base_location` for Iceberg tables. Instead, dbt, by default, writes your tables within a `_dbt/{SCHEMA_NAME}/{TABLE_NAME}` prefix to ensure easier object-store observability and auditability.
```

```

</File>

The following configurations are supported: 

| Field | Type | Required | Description | Sample input |
|-------|------|----------|-------------|-------|
|Table Format | String | Yes | Configures the object table format. | `iceberg` |
|External volume | String | Yes | Specifies the identifier (name) of the external volume where the Iceberg table stores its metadata files and data in Parquet format. If you don’t specify this parameter, the Iceberg table defaults to the external volume for the schema, database, or account. The schema takes precedence over the database, and the database takes precedence over the account.| `my_s3_bucket` |
|Base location Subpath |String | No | Defines the directory path for Snowflake to write the table data and metadata files. If you change this after creating the table, it will write a new table but leave the existing one. | `jaffle_marketing_folder` |

### Example configuration

To configure an Iceberg table materialization in dbt, refer to the example configuration:

<File name='models/<modelname>.sql'>

```sql

{{
  config(
    materialized = "table",
    table_format="iceberg",
    external_volume="s3_iceberg_snow",

>     base_location_subpath="jaffle_shop_folder",

Author: ```suggestion
```

  )
}}

select * from {{ ref('raw_orders') }}

```

</File>

### Base location 

If you were to create an Iceberg table in the Snowflake query editor, you would be required to provide a `base_location` as an input. dbt abstracts away from that. The default behavior in dbt is to provide a `base_location` string that follows this convention:

`_dbt/{SCHEMA_NAME}/{MODEL_NAME} `

#### Base Location Subpath

> We recommend using dbt's auto-generated `base_location`. However, if you have a need to customize the resulting `base_location`, dbt allows users to configure a `base_location_subpath`. When specified, it will be concatenated to the end of the previously described pattern for `base_location` string generation.

Author: ```suggestion
We recommend using dbt's auto-generated `base_location`. However, if you need to customize the resulting `base_location`, dbt allows users to configure a `base_location_subpath`. When specified, the subpath concatenates to the end of the previously described pattern for `base_location` string generation.
```

> For example, `config(base_location_subpath="prod")` will generate a `base_location` of the form `_dbt/{SCHEMA_NAME}/{MODEL_NAME}`

Author: ```suggestion
For example, `config(base_location_subpath="prod")` will generate a `base_location` of the form `_dbt/{SCHEMA_NAME}/{MODEL_NAME}`.
```

> A theoretical (not recommended) use case: re-using an `EXTERNAL VOLUME` while still maintaining isolation  across development and production environments. This is not recommended as storage permissions should configured on the external volume and underlying storage, not paths that can be modified by any analytics engineer.

Author: ```suggestion
A theoretical (but not recommended) use case is re-using an `EXTERNAL VOLUME` while maintaining isolation across development and production environments. We recommend against this as storage permissions should configured on the external volume and underlying storage, not paths that any analytics engineer can modify.
```
